## Feature Selection


Automatic feature selection techniques that you can use to prepare your machine learning data in Python with scikit-learn are:

1. Univariate Selection
2. Recursive Feature Elimination
3. Principle Component Analysis
4. Feature Importance

Feature selection is a process where you automatically select those features in your data that contribute most to the prediction variable or output in which you are interested. Having irrelevant features in your data can decrease the accuracy of many models, especially linear algorithms like linear and logistic regression. Three benefits of performing feature selection before modeling your data are:

 - Reduces Overfitting: Less redundant data means less opportunity to make decisions
based on noise.
 - Improves Accuracy: Less misleading data means modeling accuracy improves.
 - Reduces Training Time: Less data means that algorithms train faster.

You can learn more about feature selection with scikit-learn in the article [Feature selection](http://scikit-learn.org/stable/modules/feature_selection.html).
Each feature selection recipes will use the Pima Indians onset of diabetes dataset.

Statistical tests can be used to select those features that have the strongest 
relationship with the output variable. The scikit-learn library provides the 
[SelectKBest class](http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html) that can be used
with a suite of different statistical tests to select a specific number of features. 
The example below uses the chi-squared (chi) statistical test for non-negative features 
to select 4 of the best features from the Pima Indians onset of diabetes dataset.

In [1]:
# Feature Extraction with Univariate Statistical Tests (Chi-squared for classification)
from pandas import read_csv
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
# Load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
# Feature extraction
test = SelectKBest(score_func=chi2, k=4)
fit = test.fit(X, Y)
# Summarize scores
set_printoptions(precision=3)
print(fit.scores_)
features = fit.transform(X)
# Summarize selected features
print(features[0:5,:])

[ 111.52  1411.887   17.605   53.108 2175.565  127.669    5.393  181.304]
[[148.    0.   33.6  50. ]
 [ 85.    0.   26.6  31. ]
 [183.    0.   23.3  32. ]
 [ 89.   94.   28.1  21. ]
 [137.  168.   43.1  33. ]]


### Recursive Feature Elimination


The Recursive Feature Elimination (or RFE) works by recursively removing attributes and
building a model on those attributes that remain. It uses the model accuracy to identify which attributes (and combination of attributes) contribute the most to predicting the target attribute.
You can learn more about the [RFE class](http://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html) in the scikit-learn documentation. The example below
uses RFE with the logistic regression algorithm to select the top 3 features. The choice of
algorithm does not matter too much as long as it is skillful and consistent.

In [34]:
# Feature extraction wtih RFE
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
# Load Data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
# Feature extraction
##########################################
model = LogisticRegression(solver='liblinear') 
# In this we need a model for recurrently select the feature and test the dataset.
rfe = RFE(model, n_features_to_select = 3, step = 1)
fit = rfe.fit(X,Y)
###########################################
# Summarizing the selected features
print("Num Features: %d:" % rfe.n_features_)
print("Selected Features:%s" % rfe.support_)
print("Feature Ranking:%s" % rfe.ranking_)

Num Features: 3:
Selected Features:[ True False False False False  True  True False]
Feature Ranking:[1 2 3 5 6 1 1 4]


### Principal Component Analysis


Principal Component Analysis (or PCA) uses linear algebra to transform the dataset into a
compressed form. Generally this is called a data reduction technique. A property of PCA is that
you can choose the number of dimensions or principal components in the transformed result. In
the example below, we use PCA and select 3 principal components. Learn more about the PCA
class in scikit-learn by reviewing the [API](http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html).

In [45]:
# Feature Extraction with PCA
from pandas import read_csv
from sklearn.decomposition import PCA
# Load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
# Feature extraction
pca = PCA(n_components=3)
fit = pca.fit(X)
# Summarize components
print("Explained Variance:%s" % fit.explained_variance_ratio_)
print(fit.components_)

Explained Variance:[0.889 0.062 0.026]
[[-2.022e-03  9.781e-02  1.609e-02  6.076e-02  9.931e-01  1.401e-02
   5.372e-04 -3.565e-03]
 [-2.265e-02 -9.722e-01 -1.419e-01  5.786e-02  9.463e-02 -4.697e-02
  -8.168e-04 -1.402e-01]
 [-2.246e-02  1.434e-01 -9.225e-01 -3.070e-01  2.098e-02 -1.324e-01
  -6.400e-04 -1.255e-01]]


You can see that the transformed dataset (3 principal components) bare little resemblance
to the source data.

### Feature Importance


Bagged decision trees like Random Forest and Extra Trees can be used to estimate the importance
of features. In the example below we construct a ExtraTreesClassifier classifier for the Pima
Indians onset of diabetes dataset. You can learn more about the ExtraTreesClassifier class
in the scikit-learn [API](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesClassifier).

In [3]:
# Feature Importance with Extra Trees Classifier
from pandas import read_csv
from sklearn.ensemble import ExtraTreesClassifier
# Load data
filename = 'pima-indians-diabetes.data.csv'
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
# Feature extraction
model = ExtraTreesClassifier()
model.fit(X, Y)
print(model.feature_importances_)

[0.11216762 0.23628393 0.09860872 0.07832742 0.07531641 0.14023285
 0.1163603  0.14270276]


You can see that we are given an importance score for each attribute where the larger the
score, the more important the attribute. The scores suggest at the importance of plas 
 (0.23628393), mass (0.14023285) and age (0.14270276).